hystrix/hystrix.go

```diff
@@ -118,51 +97,89 @@ func Go(name string, run runFunc, fallback fallbackFunc) chan error {
 		// When requests slow down but the incoming rate of requests stays the same, you have to
 		// run more at a time to keep up. By controlling concurrency during these situations, you can
 		// shed load which accumulates due to the increasing ratio of active commands to incoming requests.
-		cmd.Lock()
+
 		select {
-		case cmd.ticket = <-circuit.executorPool.Tickets:
-			ticketChecked = true
-			ticketCond.Signal()
-			cmd.Unlock()
+		case t := <-circuit.executorPool.Tickets:
+			cmd.setTicket(t)
+
 		default:
-			ticketChecked = true
-			ticketCond.Signal()
-			cmd.Unlock()
-			returnOnce.Do(func() {
-				returnTicket()
+			select {
+			case t := <-circuit.executorPool.WaitingTicket:
+				cmd.reportEvent("queued")
```

Comment:
adding a new event type is problematic here. i agree that we should track the rate at which executions are queued, but there is a current assumption being made about the event list for an execution which no longer holds true here.

currently the `events []string` field of a command is assumed to contain data in a format of:

```
[ success|failure|rejected|short-circuit|timeout, fallback-success|fallback-failure ]
```

for example, an events slice containing `[ failure, fallback-success ]` indicates the execution failed but the fallback did not.  changing this to `[ queued, failure, fallback-success ]` makes sense (execution was queued, then failed, then fell back successfully) but other parts of the code assume the first element in the list indicates the run function's result.  this changes that and will break stats reporting as well as closing a circuit after a success.

in order to add this queued event, you'll need to change `CircuitBreaker.ReportEvent` and `metricExchange.IncrementMetrics` to account for this.

Reply:
Nice catch, I have incorporated the change by using a map structure to pass around the events.